extensions/amp-story-shopping/0.1/amp-story-shopping-config.js

```diff
@@ -14,44 +12,53 @@ import {
  *  items: !Array<!ShoppingConfigDataDef>,
  * }}
  */
-let ShoppingConfigDef;
+let ShoppingConfigResponseDef;
 
-export class AmpStoryShoppingConfig extends AMP.BaseElement {
-  /** @param {!AmpElement} element */
-  constructor(element) {
-    super(element);
-    /** @private @const {?../../amp-story/1.0/amp-story-store-service.AmpStoryStoreService} */
-    this.storeService_ = null;
-  }
+/** @typedef {!Object<string, !ShoppingConfigDataDef> */
+export let KeyedShoppingConfigDef;
 
-  /**
-   * Keys product data to product-ids and adds them to the store service.
-   * @param {!ShoppingConfigDef} shoppingConfig
-   * @private
-   */
-  addShoppingDataFromConfig_(shoppingConfig) {
-    const productIDtoProduct = {};
-    for (const item of shoppingConfig['items']) {
-      productIDtoProduct[item['product-tag-id']] = item;
-    }
-    this.storeService_.dispatch(Action.ADD_SHOPPING_DATA, productIDtoProduct);
+/**
+ * Gets Shopping config from an <amp-story-page> element.
+ * The config is validated and keyed by 'product-tag-id'.
+ * @param {!Element} pageElement <amp-story-page>
+ * @return {!Promise<!KeyedShoppingConfigDef>}
+ */
+export function getShoppingConfig(pageElement) {
+  const element = pageElement.querySelector('amp-story-shopping-config');
+  return getElementConfig(element).then((config) => {
     //TODO(#36412): Add call to validate config here.
-  }
+    return keyByProductTagId(config);
+  });
+}
 
-  /** @override */
-  buildCallback() {
-    super.buildCallback();
-    return Promise.all([
-      Services.storyStoreServiceForOrNull(this.win),
-      getElementConfig(this.element),
-    ]).then(([storeService, storyConfig]) => {
-      this.storeService_ = storeService;
-      this.addShoppingDataFromConfig_(storyConfig);
-    });
+/**
+ * @param {!ShoppingConfigResponseDef} config
+ * @return {!KeyedShoppingConfigDef}
+ */
+function keyByProductTagId(config) {
+  const keyed = {};
+  for (const item of config['items']) {
+    keyed[item['product-tag-id']] = item;
   }
+  return keyed;
+}
 
-  /** @override */
-  isLayoutSupported(layout) {
-    return layout === Layout_Enum.NODISPLAY;
-  }
+/**
+ * @param {!Element} pageElement
+ * @param {!KeyedShoppingConfigDef} config
+ * @return {!Promise<!ShoppingConfigResponseDef>}
+ */
+export function storeShoppingConfig(pageElement, config) {
+  const win = pageElement.ownerDocument.defaultView;
+  return Services.storyStoreServiceForOrNull(win).then((storeService) => {
+    // TODO(wg-stories): Scope shopping data to page id:
```

Comment:
Can you explain in a bit more detail about this TODO?

Reply:
Sure. The `ADD_SHOPPING_DATA` action merges the passed object with the existing object:

```js
      const updatedShoppingData = {
        ...state[StateProperty.SHOPPING_DATA],
        ...data,
      };
```

Since the passed object is directly keyed by `product-tag-id`, this means that data can be corrupted by collisions:

```js
// page-1
{"foo": "bar", "baz": "qux"}
// page-2
{"abc": "xyz", "baz": "345"}
// result (bad, "baz" is overridden)
{"foo": "bar", "abc": "xyz", "baz": "345"} 
```

Instead, by adding another level keyed by page id we avoid this conflict:
```js
{
  "page-1": {"foo": "bar", "baz": "qux"},
  "page-2": {"abc": "xyz", "baz": "345"},
}
```

This issue is currently masked by the consumer in `AmpStoryShoppingConfig` tag due to this condition, causing the tag to render only once:

```js
    if (this.hasAppendedInnerShoppingTagEl_ || !this.tagData_) {
      return;
    }
```

However, this would be a problem when the data is consumed later on, either by the necessity of updates or a different component reading later.

Reply:
Alright, if it's intentional then I'll let it be.